grid/app/main/cycles/cycle_manager.py

```diff
@@ -185,21 +185,20 @@ def complete_cycle(self, cycle_id: str):
         )
         logging.info("# of diffs: %d" % completed_cycles_num)
 
-        min_worker = server_config.get("min_worker", 3)
-        max_worker = server_config.get("max_worker", 3)
-        received_diffs_exceeds_min_worker = completed_cycles_num >= min_worker
-        received_diffs_exceeds_max_worker = completed_cycles_num >= max_worker
-        cycle_ended = True  # check cycle.cycle_time (but we should probably track cycle startime too)
-
-        # Hmm, I don't think there should be such connection between ready_to_average, max_workers, and received_diffs
-        # I thought max_workers just caps total number of simultaneous workers
-        # 'cycle end' condition should probably depend on cycle_length regardless of number of actual received diffs
-        # another 'cycle end' condition can be based on min_diffs
+        min_diffs = server_config.get("min_diffs", 1)
+        max_diffs = server_config.get("max_diffs", 999999999999)
+
+        cycle_ended = (
+            True
+            if (datetime.now() >= cycle.end or completed_cycles_num >= max_diffs)
+            else False
+        )
+
         ready_to_average = (
             True
             if (
-                (received_diffs_exceeds_max_worker or cycle_ended)
-                and received_diffs_exceeds_min_worker
+                (cycle_ended and completed_cycles_num >= min_diffs)
+                or (min_diffs is not None and completed_cycles_num >= min_diffs)
```

Comment:
Hmm, 2nd condition can be true regardless of `cycle_ended`, i.e. max_diffs and cycle_length will be ignored.

Reply:
I think this is what we finally arrived to in slack discussion:

```
hit_diffs_limit = completed_cycles_num >= max_diffs if max_diffs is not None else False
hit_time_limit = now() >= cycle.end if cycle.end is not None else False
no_limits = max_diffs is None and cycle.end is None
has_enough_diffs = completed_cycles_num >= min_diffs if min_diffs is not None else True

ready_to_average = has_enough_diffs and (no_limits or hit_diffs_limit or hit_time_limit)
```